### PR TITLE
Remove co dependency

### LINKED ISSUE #557: Drop co dependency
This has already been discussed in #533 and #415 . Quoting @yanickrochon:
 > Since  co  has been Promise-based for a while now, I think Koa should drop it as dependency and move to a Promise solution. The advantage of using promises is quite evident and should not have a negative impact in most cases. Also, old middlewares can use  co.wrap  if they need to be re-used, until they migrate to be 2.0 compatible. To me, since there is a solution that not only works but is clean means that there's no problem in dropping  co  in favour of Promise or async/await (using a transpiler for now).
> 
 > Furthermore, Generators have always been "hacky". It does not promote good usage of the design pattern :)

This would also allow us to use Bluebird.coroutine instead of co if we want.
Would submit PR

### PR DESCRIPTION
**Pull request by felixfbecker**
Closes #557

### CODE REVIEW

**Comment by felixfbecker** [comment]
Unfortunately it is hard for me to test because this module relies on makefiles and `make` isn't available on Windows. It would be nice if the tests just worked without `make`.

**Comment by stojanovic** [comment]
@felixfbecker I can send PR tomorrow to add everything from Makefile to npm scripts, so you'll be able to run tests and lint without `make`.

**Comment by felixfbecker** [comment]
Did it by running mocha and eslint manually.

**Comment by stojanovic** [comment]
We just need that in npm scripts :)

**Comment by yanickrochon** [comment]
Before this PR is merged, all documentations should be updated to reflect the change :)

**Comment by jonathanong** [comment]
anyone object to this?

**Comment by tejasmanohar** [comment]
:+1: from me. we should put up a migrating from koa 1.x.x -> 2.x.x / major changes doc, right? @jonathanong - If so, I can PR it later this week

**Comment by tj** [comment]
:+1:

File: test/application/use.js [hunk 1]
```diff
@@ -9,22 +9,25 @@ describe('app.use(fn)', function(){
     const app = new Koa();
     const calls = [];
 
-    app.use(function *(ctx, next){
+    app.use(function(ctx, next){
       calls.push(1);
-      yield next();
-      calls.push(6);
+      next().then(function(){
```
Comment (gyson) [inline test/application/use.js]: `return next()`

File: test/application/use.js [hunk 2]
```diff
@@ -9,22 +9,25 @@ describe('app.use(fn)', function(){
...
+        calls.push(6);
+      });
     });
 
-    app.use(function *(ctx, next){
+    app.use(function(ctx, next){
       calls.push(2);
-      yield next();
-      calls.push(5);
+      next().then(function(){
```
Comment (gyson) [inline test/application/use.js]: `return next()`

File: test/application/use.js [hunk 3]
```diff
@@ -9,22 +9,25 @@ describe('app.use(fn)', function(){
...
+        calls.push(5);
+      });
     });
 
-    app.use(function *(ctx, next){
+    app.use(function(ctx, next){
       calls.push(3);
-      yield next();
-      calls.push(4);
+      next().then(function(){
```
Comment (gyson) [inline test/application/use.js]: `return next()`

**Comment by felixfbecker** [comment]
@gyson good find. Kind of confusing though that all tests pass even though I didn't do this?!

**Comment by gyson** [comment]
@felixfbecker because this test case is very simple and it just accidentally passed even w/o `return`.

Tests would fail if we test `.catch` or delay some `next()` call.

For example

```js
const Koa = require('koa')

const app = new Koa();
const calls = [];

app.use(function(ctx, next){
  calls.push(1);
  return next().catch(function(){
    calls.push(4); // w/o `return`, this wont be called
  });
});

app.use(function(ctx, next){
  calls.push(2);
  return next().then(function(){
    calls.push(3);
    throw new Error()
  });
});
```

File: lib/application.js [hunk 4]
```diff
@@ -93,7 +92,9 @@ module.exports = class Application extends Emitter {
   use(fn) {
     debug('use %s', fn._name || fn.name || '-');
     if (typeof fn !== 'function') throw new TypeError('middleware must be a function!');
-    if (isGeneratorFunction(fn)) fn = co.wrap(fn);
+    if (isGeneratorFunction(fn)) {
+      throw new Error('Support for generators has been removed. Use Promises or wrap your generator with co.wrap');
```
Comment (jonathanong) [inline lib/application.js]: should be a `TypeError` :)

Comment (jonathanong) [inline lib/application.js]: i'll fix, squash, and merge

Comment (felixfbecker) [inline lib/application.js]: You're right of course.

**Pull request closed without merging**